cardano-cli/src/Cardano/CLI/Shelley/Run/Genesis.hs

```diff
@@ -683,38 +693,43 @@ runGenesisCreateStaked (GenesisDir rootdir)
   forM_ (zip [ 1 .. bulkPoolCredFiles ] bulkSlices) $
     uncurry (writeBulkPoolCredentials pooldir)
 
-  forM_ [ 1 .. genNumStDelegs ] $ \index ->
-    createDelegatorCredentials stdeldir index
-
-  delegations :: [Delegation] <-
-    -- Distribute M delegates across N pools:
-    forM [ (pool, delegIx)
-         | (pool, poolIx) <- zip pools [1 ..]
-         , delegIxLocal <- [ 1 .. delegsPerPool ] ++
-                           -- Add all remaining delegates to the last pool:
-                           if delegsRemaining /= 0 && poolIx == genNumPools
-                           then [ delegsPerPool + 1 .. delegsPerPool + delegsRemaining ]
-                           else []
-         , let delegIx = delegIxLocal + delegsPerPool * (poolIx - 1)] $
-      uncurry (computeDelegation network stdeldir)
+  let (delegsPerPool, delegsRemaining) = divMod genNumStDelegs genNumPools
+      delegsForPool poolIx = if delegsRemaining /= 0 && poolIx == genNumPools
+        then delegsPerPool
+        else delegsPerPool + delegsRemaining
+      distribution = [pool | (pool, poolIx) <- zip pools [1 ..], _ <- [1 .. delegsForPool poolIx]]
+
+  -- Distribute M delegates across N pools:
+  delegations <- liftIO $ Lazy.forM distribution $ computeDelegation network
+
+  -- NOTE The following code which reads from the same file 'delegations.jsonl' multiple times
+  -- looks like duplication, but it is not.  The file is read lazily, and it is important that
+  -- they be read multiple times because the code is streaming and reading the file multiple
+  -- times ensures that any data structures that are created as a result of the read is not
+  -- retained in memory.
+
+  let numDelegations = length delegations
```

Comment:
This is suspicious, calling `length` could destroy the laziness. We could get `numDelegations` in this list comprehension: `distribution = [pool | (pool, poolIx) <- zip pools [1 ..], _ <- [1 .. delegsForPool poolIx]]` if we multiply the number of pools by the number of delegations per pool

Reply:
Yes, calling `length` destroys laziness, but I've been in discussion with @kosyrevSerge and he is happy to pay the memory cost in this case.

Note, `length` is not the only thing that "destroys laziness".  Using `delegations` more than once does the same thing.  Previously, `delegations` was written to a temporary file and the temporarily file was read multiple times to preserve laziness, but @kosyrevSerge didn't want to pay the serialisation cost of that.

Reply:
The comment above is out of date however and I've deleted it.